standard/expressions.md

```diff
@@ -2906,6 +2907,77 @@ A *default_value_expression* is a constant expression ([§11.21](expressions.md#
+    ;
+
+stackalloc_initializer_elements
+     : '{' stackalloc_initializer_element_list '}'
+     ;
+
+stackalloc_initializer_element_list
+     : stackalloc_element_initializer (',' stackalloc_element_initializer)* ','?
+     ;
+    
+stackalloc_element_initializer
+    : expression
+    ;
+```
+
+The *unmanaged_type* ([§8.8](types.md#88-unmanaged-types)) indicates the type of the items that will be stored in the newly allocated location, and the *expression* indicates the number of these items. Taken together, these specify the required allocation size. As the size of a stack allocation cannot be negative, it is a compile-time error to specify the number of items as a *constant_expression* that evaluates to a negative value.
+
+If *unmanaged_type* is omitted, it is inferred from the corresponding *stackalloc_initializer_elements*. If *expression* is omitted from *stackalloc_initializer*, it is inferred to be the number of *stackalloc_element_initializer*s in the corresponding *stackalloc_initializer_elements*.
+
+When a *stackalloc_initializer* includes both *expression* and *stackalloc_initializer_elements*, the *expression* shall be a *constant_expression* and the number of elements in that *stackalloc_initializer_elements* shall match the value of *expression*.
+
+A stack allocation initializer of the form `stackalloc T[E]` requires `T` to be an *unmanaged_type* and `E` to be an expression implicitly convertible to type `int`. The operator allocates `E * sizeof(T)` bytes from the call stack. The result is a pointer, of type `T*`, to the newly allocated block. For use in safe contexts, a *stackalloc_initializer* has an implicit conversion from `T*` to `Span<T>`. As pointer contexts require unsafe code, see §stack-allocation for more information.
+
+If `E` is a negative value, then the behavior is undefined. If `E` is zero, then no allocation is made, and the value returned is implementation-defined. If there is not enough memory available to allocate a block of the given size, a `System.StackOverflowException`  is thrown.
+
+When *stackalloc_initializer_elements* is present, the *stackalloc_initializer_element_list* shall consist of a sequence of expressions, each having an implicit conversion to *unmanaged_type* ([§10.2](conversions.md#102-implicit-conversions)). The expressions initialize elements in the allocated memory in increasing order, starting with the element at index zero. In the absence of a *stackalloc_initializer_elements*, the content of the newly allocated memory is undefined.
+
+Access via an instance of `System.Span<T>` to the elements of an allocated block is range checked.
+
+Stack allocation initializers are not permitted in `catch` or `finally` blocks ([§12.11](statements.md#1211-the-try-statement)).
+
+> *Note*: Stack allocation initializers are allowed in `async` methods, but their return value can't be assigned. Neither pointers nor `ref struct` types, like `Span<T>` are allowed in `async` methods. *end note*
```

Comment:
This is because the following is legal (if not particularly useful):

```csharp
public async Task<int> M4()
{
    _ = stackalloc int[3];
    await Task.Delay(1000);
    return 5;
}
```

But, this isn't legal:

public async Task<int> M4()
{
    Span<int> storage = stackalloc int[3]; // error can't declare a ref struct in an async method
    await Task.Delay(1000);
    return 5;
}

Reply:
Yes, this is legal:

```csharp
async Task<int> M4()
{
    int r = M5(stackalloc int[3]);
    await Task.Delay(1000);
    return 5;
}

int M5(Span<int> arg)
{
    return arg.Length;
}
```

Reply:
Perhaps (in a follow-up PR) it would be worth given that as an example. Otherwise it looks somewhat useless...

Reply:
Is it legal?

According to text a *stackalloc_initializer* is only valid as a *local_variable_initializer*. It is not a general expression in a similar way that *array_initializer* is not, both are included in the grammar in the places they are allowed.

My other comment *removes* it from *primary_no_array_creation_expression* precisely for this reason. (If it does need to be an expression then we need to check for any grammar gotchas as there would be two routes to it from *local_variable_initializer*)

Reply:
Decision from 5/17 meeting: `stackalloc` is an expression. It should not be restricted to a *local_variable_initializer*.